lib/async_hooks.js

```diff
@@ -0,0 +1,486 @@
+      hook_fields = tmp_async_hook_fields;
+    }
+
+    // Each hook is only allowed to be added once.
+    if (hooks_array.includes(this))
+      return;
+
+    // createHook() has already enforced that the callbacks are all functions,
+    // so here simply increment the count of whether each callbacks exists or
+    // not.
+    hook_fields[kInit] += +!!this[init_symbol];
+    hook_fields[kBefore] += +!!this[before_symbol];
+    hook_fields[kAfter] += +!!this[after_symbol];
+    hook_fields[kDestroy] += +!!this[destroy_symbol];
+    hooks_array.push(this);
+    return this;
+  }
+
+  disable() {
+    // Use local reference in case the pointer to the array needs to change.
+    var hooks_array = active_hooks_array;
+    var hook_fields = async_hook_fields;
+
+    if (processing_hooks) {
+      if (tmp_active_hooks_array === null)
+        storeActiveHooks();
+      hooks_array = tmp_active_hooks_array;
+      hook_fields = tmp_async_hook_fields;
+    }
+
+    const index = hooks_array.indexOf(this);
+    if (index === -1)
+      return;
+
+    hook_fields[kInit] -= +!!this[init_symbol];
+    hook_fields[kBefore] -= +!!this[before_symbol];
+    hook_fields[kAfter] -= +!!this[after_symbol];
+    hook_fields[kDestroy] -= +!!this[destroy_symbol];
+    hooks_array.splice(index, 1);
+    return this;
+  }
+}
+
+
+function storeActiveHooks() {
+  tmp_active_hooks_array = active_hooks_array.slice();
+  // Don't want to make the assumption that kInit to kDestroy are indexes 0 to
+  // 4. So do this the long way.
+  tmp_async_hook_fields = [];
+  tmp_async_hook_fields[kInit] = async_hook_fields[kInit];
+  tmp_async_hook_fields[kBefore] = async_hook_fields[kBefore];
+  tmp_async_hook_fields[kAfter] = async_hook_fields[kAfter];
+  tmp_async_hook_fields[kDestroy] = async_hook_fields[kDestroy];
+}
+
+
+// Then restore the correct hooks array in case any hooks were added/removed
+// during hook callback execution.
+function restoreTmpHooks() {
+  active_hooks_array = tmp_active_hooks_array;
+  async_hook_fields[kInit] = tmp_async_hook_fields[kInit];
+  async_hook_fields[kBefore] = tmp_async_hook_fields[kBefore];
+  async_hook_fields[kAfter] = tmp_async_hook_fields[kAfter];
+  async_hook_fields[kDestroy] = tmp_async_hook_fields[kDestroy];
+
+  tmp_active_hooks_array = null;
+  tmp_async_hook_fields = null;
+}
+
+
+function createHook(fns) {
+  return new AsyncHook(fns);
+}
+
+
+function currentId() {
+  return async_uid_fields[kCurrentAsyncId];
+}
+
+
+function triggerId() {
+  return async_uid_fields[kCurrentTriggerId];
+}
+
+
+// Embedder API //
+
+class AsyncEvent {
+  constructor(type, triggerId) {
+    this[async_id_symbol] = ++async_uid_fields[kAsyncUidCntr];
+    // Do this before early return to null any potential kInitTriggerId.
+    if (triggerId === undefined)
+      triggerId = initTriggerId();
+    // If a triggerId was passed, any kInitTriggerId still must be null'd.
+    else
+      async_uid_fields[kInitTriggerId] = 0;
```

Comment:
oops. usually am more careful about that. in this case you're referring to the comment above the `else` correct?